csrc/preseg_passes/allocation_order_inference.cpp

```diff
@@ -52,54 +57,70 @@ int64_t countNonTrivialIterDomains(const TensorView* tv) {
 //   (sharp-edge 1: if target_logical_domain[i] is a reduction and is not
 //   mapped, we keep the reduction iter domain in the original position.) Push
 //   the front of unmapped_id_vec to the end of target allocation domain, if
-//   unmapped_id_vec isn't empty yet; Otherwise, push the frnot of mapped_ids at
-//   the end of target allocation domain.
+//   unmapped_id_vec isn't empty yet; Otherwise, push the rfront of mapped_ids
+//   at the end of target allocation domain.
 //
 // Note: we could be using a simplified logic below,
 // See issue https://github.com/NVIDIA/Fuser/issues/2202
 // 1. we project iter domains from targets' logical domain which has an exact
 // map to ref's allocation domain.
-//   mapped_ids {ir5[i1], iS7[i2]}
+//   mapped_ids {iS7[i2], iS5[i1]}
 // 2. remove all projected iter domains from target's rfactor
 // domain:
 //   unmapped_ids {iS3[i3], iS4[i4], iS6[i5], ir8[1]}
-// 3. append mapped_ids at the end of unmapped_id_vec.
+// 3. append reversed mapped_ids at the end of unmapped_id_vec.
 //   target_alloc_domain
-//   {iS3[i3], iS4[i4], iS6[i5], ir8[1], ir5[i1], iS7[i2]}
+//   {iS3[i3], iS4[i4], iS6[i5], ir8[1], iS5[i1], iS7[i2]}
 void mapAllocationDomain(
     const IdModel& id_model,
     const TensorView* ref,
     TensorView* target) {
-  const ValGraph& val_graph = id_model.idGraph(IdMappingMode::EXACT);
+  const ValGraph& exact_graph = id_model.idGraph(IdMappingMode::EXACT);
 
   std::vector<IterDomain*> ref_alloc_domain = ref->getMaybeAllocationDomain();
+  // reverse ref_alloc_domain, so a range based loop would iterate through from
+  // fast to slow dimensions
+  std::reverse(ref_alloc_domain.begin(), ref_alloc_domain.end());
   const std::vector<IterDomain*>& target_logical_domain =
       target->getLogicalDomain();
 
   // map target logical domain into ref's allocation domain
   nvfuser::VectorOfUniqueEntries<IterDomain*> mapped_ids;
```

Comment:
Can you remind me why a vector isn't enough?

Reply:
```
142   std::vector<IterDomain*> unmapped_ids = target_logical_domain;
143   auto unmapped_ids_vec_end = std::remove_if(
144       unmapped_ids.begin(), unmapped_ids.end(), [&mapped_ids](IterDomain* it) {
145         return mapped_ids.has(it) || it->isReduction();
146       });
```

This is just for a couple use of `mapped_ids.has()`.

Reply:
I see. How comfortable are you with the assumption that `IterDomain*`s are unique? I'm a bit worried about VectorOfUniqueEntries *quietly* ignore duplicated elements.

Reply:
IIUC, multiple entries in a single TensorDomain is called `self_mapping`. The propagation pass currently don't handle the case at all.

We'll skip the source/target when self mapping is detected:
https://github.com/NVIDIA/Fuser/blob/88f6a6295fc652f10b4c86416744c6b69eb275cc/csrc/preseg_passes/allocation_order_inference.cpp#L227
https://github.com/NVIDIA/Fuser/blob/88f6a6295fc652f10b4c86416744c6b69eb275cc/csrc/preseg_passes/allocation_order_inference.cpp#L241